src/services/data-targets/fiware-data-target.service.ts

```diff
@@ -9,9 +10,53 @@ import { DataTargetSendStatus } from "@interfaces/data-target-send-status.interf
 import { FiwareDataTargetConfiguration } from "@interfaces/fiware-data-target-configuration.interface";
 import { BaseDataTargetService } from "@services/data-targets/base-data-target.service";
 
+type TokenEndpointResponse = {
+    access_token: string,
+    expires_in: number,
+}
+
+@Injectable()
+export class AuthenticationTokenProvider {
+
+    private readonly logger = new Logger(AuthenticationTokenProvider.name);
+
+    constructor(private httpService: HttpService, @Inject(CACHE_MANAGER) private cacheManager: Cache) {
+    }
+
+    async getToken(config: FiwareDataTargetConfiguration): Promise<string> {
+        const key = config.clientId + config.clientSecret;
+        const token = await this.cacheManager.get<string>(key)
```

Comment:
Caching can be nice, if handled properly. The token isn't invalided if the request fails. This can soft-lock the data target until the token expires.

Reply:
Do you want to remove it only from cache or also ask for another and try again?

Reply:
Retry makes for a better user flow. If a request fails, instead of potentially losing data, then perform a retry with a fresh token if and only if it was due to the token being expired/invalid.

Reply:
Done